pkg/controllers/management/clusterprovisioner/helm_cleaner.go

```diff
@@ -0,0 +1,296 @@
+		"istio-system",               // Istio and Sub-charts
+		"cattle-istio-system",        // Kiali
+		"cattle-logging-system",      // Logging
+		"cattle-windows-gmsa-system", // Windows GMSA
+		"cattle-sriov-system",        // Sriov
+		"cattle-ui-plugin-system",    // UI Plugin System
+	}
+)
+
+// EmptyHelmDriverName is a placeholder for the empty Helm driver.
+const EmptyHelmDriverName = ""
+
+func newClientGetter(k8sClient kubernetes.Interface, restConfig rest.Config) *wrangler.SimpleRESTClientGetter {
+	return &wrangler.SimpleRESTClientGetter{
+		ClientConfig:    &clientcmd.DefaultClientConfig,
+		RESTConfig:      &restConfig,
+		CachedDiscovery: memory.NewMemCacheClient(k8sClient.Discovery()),
+		RESTMapper:      meta.NewDefaultRESTMapper(nil),
+	}
+}
+
+func (p *Provisioner) cleanupHelmReleases(cluster *v3.Cluster) error {
+	clusterManager := p.clusterManager
+	userContext, err := clusterManager.UserContextNoControllers(cluster.Name)
+	if err != nil {
+		return errors.Wrapf(err, "[cleanupHelmReleases] failed to obtain the Kubernetes client instance")
+	}
+
+	client := userContext.K8sClient
+	restConfig := userContext.RESTConfig
+
+	clientGetter := newClientGetter(client, restConfig)
+
+	for _, namespace := range FeatureAppNS {
+		// TODO check if there's a way of doing this using a more API-based approach
+		actionConfig := &action.Configuration{}
+		if err = actionConfig.Init(clientGetter, namespace, EmptyHelmDriverName, logrus.Debugf); err != nil {
+			return errors.Wrapf(err, "[cleanupHelmReleases] failed to create ActionConfiguration instance for Helm")
+		}
+
+		listAction := action.NewList(actionConfig)
+		releases, err := listAction.Run()
+		if err != nil {
+			return errors.Wrapf(err, "[cleanupHelmReleases] failed to list Helm releases for namespace %v", namespace)
+		}
+
+		for _, helmRelease := range releases {
+			lastRelease, err := actionConfig.Releases.Last(helmRelease.Name)
+			if err != nil {
+				logrus.Errorf("[cleanupHelmReleases] failed to find latest release version for release %v", helmRelease.Name)
+				// If this fails, something went wrong. Skip to the next release.
+				continue
+			}
+
+			// TODO consume the function from helm-mapkubeapis once that is merged in
+			replaced, modifiedManifest, err := ReplaceManifestData(apiMappings, lastRelease.Manifest, cluster.Status.Version.GitVersion)
+			if err != nil {
+				// If this fails, it probably means we don't have adequate write permissions
+				return errors.Wrapf(err, "[cleanupHelmReleases] failed to replace deprecated/removed APIs on cluster %v", cluster.Name)
+			}
+
+			if !replaced {
+				logrus.Infof("[cleanupHelmReleases] release %v in namespace %v has no deprecated or removed APIs", lastRelease.Name, namespace)
+				continue
+			}
+
+			if err := updateRelease(lastRelease, modifiedManifest, actionConfig); err != nil {
+				logrus.Errorf("[cleanupHelmReleases] failed to update release %v in namespace %v, skipping...", lastRelease.Name, lastRelease.Namespace)
```

Comment:
Yes that's a good point. My concern is that it would keep re-enqueuing and retrying for things that are unrecoverable (such as an error in the mapping file). What's your idea for such a situation?

Reply:
I could return different error types to handle unrecoverable cases vs recoverable cases. What do you think? How is this handled in Rancher's codebase currently?

Reply:
That's a good idea, I think that will work out well.

I don't know if there is a good exact replica of this. Most of the time, if we are checking the error type returned by k8s - see here for an example: https://github.com/rancher/rancher/blob/b82ee81ce21a7290c8ade98eb289c07e85c2f9c3/pkg/api/norman/customization/app/app.go#L119. 

Generally I think the strategy here is:
- Define a new type which implements the error interface
- Implement a function which allows you to do type checking
- Return the type as needed
- Check using your new function

Reply:
I'm aware of general strategies in Go, but was more curious about our established process. There are parts that we may not be able to tell whether it is retryable or not (think Helm internal errors) but for the ones we bubble up it should work.

Reply:
Outside of the general go practice and what I've linked, I don't think there's anything relevant here. We do have some logic where we retry a specific operation/piece of code until it succeeds, but I don't think that will help you here.

Reply:
Updated this!